File: swift/Sources/Wallet/IssuanceSession.swift
import Foundation

/**
 * A single issuance run driven as a state machine. One-shot; the flow runs in
 * a Task and pauses at browser-authorization / tx-code interruptions until the app resumes.
 */
public final class IssuanceSession: @unchecked Sendable {
    private let lock = NSLock()
    private var _currentState: IssuanceState = .preparing
    public var currentState: IssuanceState { locked { _currentState } }

    public let states: AsyncStream<IssuanceState>
    private let continuation: AsyncStream<IssuanceState>.Continuation
    private var task: Task<Void, Never>?
    private var authContinuation: CheckedContinuation<String, Never>?
    private var txCodeContinuation: CheckedContinuation<String, Never>?
    private var _issuer: String?
    private let flow: (IssuanceSession) async throws -> Void

    /// The credential issuer, set by the flow as soon as it is known, so a failure can be logged against it.
    var issuer: String? {
        get { locked { _issuer } }
        set { locked { _issuer = newValue } }
    }

    init(_ flow: @escaping (IssuanceSession) async throws -> Void) {
        self.flow = flow
        var c: AsyncStream<IssuanceState>.Continuation!
        self.states = AsyncStream { c = $0 }
        self.continuation = c
    }

    func launch() {
        task = Task { [self] in
            do {
                try await flow(self)
            } catch let error as IssuanceError {
                emit(.failed(error))
            } catch {
                emit(.failed(.unexpected(String(describing: error))))
            }
        }
    }

    func emit(_ state: IssuanceState) {
        locked { _currentState = state }
        continuation.yield(state)
        if state.isTerminal { continuation.finish() }
    }

    /// Pauses the flow at `.authorizationRequired` until `completeAuthorization`.
    func awaitAuthorization(_ url: String) async -> String {
        await withCheckedContinuation { cont in
            locked { authContinuation = cont } // set before emitting to avoid a resume-before-suspend race
            emit(.authorizationRequired(url))
        }
    }

    /// Pauses the flow at `.txCodeRequired` until `submitTxCode`.
    func awaitTxCode(_ txCode: TxCodeSpec?) async -> String {
        await withCheckedContinuation { cont in
            locked { txCodeContinuation = cont }
            emit(.txCodeRequired(txCode))
        }
    }

    /// Resume after the browser authorization step (auth-code flow).
    public func completeAuthorization(_ redirectUri: String) {
        let cont = locked { () -> CheckedContinuation<String, Never>? in
            let c = authContinuation; authContinuation = nil; return c
        }
        cont?.resume(returning: redirectUri)
    }

    /// Provide the transaction code when the session is `.txCodeRequired`.
    public func submitTxCode(_ code: String) {
        let cont = locked { () -> CheckedContinuation<String, Never>? in
            let c = txCodeContinuation; txCodeContinuation = nil; return c
        }
        cont?.resume(returning: code)
    }

    public func cancel() {
        task?.cancel()
        continuation.finish()
    }

    private func locked<T>(_ body: () -> T) -> T {
        lock.lock()
        defer { lock.unlock() }
        return body()
    }
}

File: kotlin/wallet/src/main/kotlin/com/hopae/eudi/wallet/IssuanceSession.kt
package com.hopae.eudi.wallet

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * A single issuance run driven as a state machine. One-shot; the flow runs in
 * the wallet scope and pauses at browser-authorization / tx-code interruptions until the app resumes.
 */
class IssuanceSession internal constructor(
    private val scope: CoroutineScope,
    private val flow: suspend IssuanceSession.() -> Unit,
) {
    private val _state = MutableStateFlow<IssuanceState>(IssuanceState.Preparing)
    val state: StateFlow<IssuanceState> = _state.asStateFlow()

    private var job: Job? = null
    private var authContinuation: CompletableDeferred<String>? = null
    private var txCodeContinuation: CompletableDeferred<String>? = null

    /** The credential issuer, set by the flow as soon as it is known, so a failure can be logged against it. */
    internal var issuer: String? = null

    internal fun launch() {
        job = scope.launch {
            try {
                flow()
            } catch (e: WalletError.Issuance) {
                _state.value = IssuanceState.Failed(e)
            } catch (e: Throwable) {
                _state.value = IssuanceState.Failed(WalletError.Issuance.Unexpected(e))
            }
        }
    }

    internal fun emit(state: IssuanceState) {
        _state.value = state
    }

    /** Pauses the flow at [IssuanceState.AuthorizationRequired] until [completeAuthorization]. */
    internal suspend fun awaitAuthorization(url: String): String {
        val deferred = CompletableDeferred<String>()
        authContinuation = deferred
        emit(IssuanceState.AuthorizationRequired(url))
        return deferred.await()
    }

    /** Pauses the flow at [IssuanceState.TxCodeRequired] until [submitTxCode]. */
    internal suspend fun awaitTxCode(txCode: TxCodeSpec?): String {
        val deferred = CompletableDeferred<String>()
        txCodeContinuation = deferred
        emit(IssuanceState.TxCodeRequired(txCode))
        return deferred.await()
    }

    /** Resume after the browser authorization step (auth-code flow). */
    fun completeAuthorization(redirectUri: String) {
        authContinuation?.complete(redirectUri) ?: error("session is not awaiting authorization")
    }

    /** Provide the transaction code when the session is [IssuanceState.TxCodeRequired]. */
    fun submitTxCode(code: String) {
        txCodeContinuation?.complete(code) ?: error("session is not awaiting a transaction code")
    }

    fun cancel() {
        job?.cancel()
    }
}
